Include SizingPeriod:WeatherFileDays in DOAS sizing

rraustad opened the pull request:
Pull request overview
---------------------
 - Follow up to #9327
 - Fixes #9332 to include SizingPeriod:WeatherFileDays in DOAS sizing

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

rraustad committed:
Include RunPeriodDesign in DOAS sizing [decent_ci_skip]

rraustad commented:
This could be included in 22.1 but I think there could be improvements. Maybe these changes are included in 22.1 and the improvements in 22.2?

nrel-bot-2 commented:
@rraustad @myoldmopar it has been 28 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @myoldmopar it has been 28 days since this pull request was last updated.

Myoldmopar commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2045,6 +2045,29 @@ namespace WeatherManager {
+                            state.dataWeatherManager->TodayOutDewPointTemp[i];
+                    }
+                }
+            }
```
For reference (pun intended), consider the readability of these two:

```
if (state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn).KindOfEnvrn ==
        DataGlobalConstants::KindOfSim::RunPeriodDesign ||
    state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn).KindOfEnvrn == DataGlobalConstants::KindOfSim::DesignDay) {
  for (int i = 0; i < state.dataWeatherManager->TodayOutDryBulbTemp.size(); ++i) {
    if (state.dataWeatherManager->TodayOutDryBulbTemp[i] >
        state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn).maxCoolingOATSizing) {
      state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn).maxCoolingOATSizing =
          state.dataWeatherManager->TodayOutDryBulbTemp[i];
      state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn).maxCoolingOADPSizing =
          state.dataWeatherManager->TodayOutDewPointTemp[i];
    }
    if (state.dataWeatherManager->TodayOutDryBulbTemp[i] <
        state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn).minHeatingOATSizing) {
      state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn).minHeatingOATSizing =
          state.dataWeatherManager->TodayOutDryBulbTemp[i];
      state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn).minHeatingOADPSizing =
          state.dataWeatherManager->TodayOutDewPointTemp[i];
    }
  }
}
```

vs

```
auto &thisEnvrn = state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn);
if (thisEnvrn.KindOfEnvrn == DataGlobalConstants::KindOfSim::RunPeriodDesign || thisEnvrn.KindOfEnvrn == DataGlobalConstants::KindOfSim::DesignDay) {
  for (int i = 0; i < state.dataWeatherManager->TodayOutDryBulbTemp.size(); ++i) {
    auto &thisDB = state.dataWeatherManager->TodayOutDryBulbTemp[i];
    auto &thisDP = state.dataWeatherManager->TodayOutDewPointTemp[i];
    if (thisDB > thisEnvrn.maxCoolingOATSizing) {
      thisEnvrn.maxCoolingOATSizing = thisDB;
      thisEnvrn.maxCoolingOADPSizing = thisDP;
    }
    if (thisDB < thisEnvrn.minHeatingOATSizing) {
      thisEnvrn.minHeatingOATSizing = thisDB;
      thisEnvrn.minHeatingOADPSizing = thisDP;
    }
  }
}
```

Just by adding a few reference variables, the logic/algorithmic part of the code becomes instantly and significantly more readable.

amirroth replied:
:rage: @Myoldmopar, you and your `auto &` for scalars! Just use `Real64` for `thisDB` and `thisDP`. Telling the compiler that you want to access these by reference just gives the compiler two more addresses to track in this function.  

PS Weather manager should be refactored to use a single 2D array (or even a 1D array) of `structs` rather than a separate 2D array for each quantity. That would also simplify these kinds of loops. :pushpin:

Myoldmopar replied:
Yep! Sorry about that. I'm so accustomed to creating references to the objects, it just gets into my muscle memory.

rraustad replied:
The last tight block does look much better.

amirroth replied:
@rraustad (and everyone). The general "rule" is:

- `auto &` (or `auto const &`) for classes. I hate `auto` in general, but using it to mean "`class` whose name I can't quite remember right now and is probably too long anyway" makes it slightly less terrible.
- The actual type for scalars and strings with no `&`. `int`, `Real64`, `bool` are all about as long as `auto` and not hard to remember. I would also use `std::string`, `std::string &` and `std::string_view` explicitly because they have different copy semantics and I am not sure the compiler can figure out which to use on its own.
- No `auto &` for arrays themselves.

Myoldmopar commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -1019,40 +1019,27 @@ namespace AirLoopHVACDOAS {
+        // int constexpr summerDesignDayTypeIndex(9);
+        // int constexpr winterDesignDayTypeIndex(10);
+        for (size_t env = 1; env <= state.dataWeatherManager->Environment.size(); ++env) {
+            switch (state.dataWeatherManager->Environment(env).KindOfEnvrn) {
```
Given that this is only acting on two possible conditions, I'm not _convinced_ that the switch block is better, but at the same time, I personally think it could add a lot of readability with a couple tweaks.  I'm curious what @amirroth would comment here.

Right now the code looks like:
```
// int constexpr summerDesignDayTypeIndex(9);
// int constexpr winterDesignDayTypeIndex(10);
for (size_t env = 1; env <= state.dataWeatherManager->Environment.size(); ++env) {
  switch (state.dataWeatherManager->Environment(env).KindOfEnvrn) {
  case DataGlobalConstants::KindOfSim::DesignDay:
  case DataGlobalConstants::KindOfSim::RunPeriodDesign: {
    if (state.dataWeatherManager->Environment(env).maxCoolingOATSizing >
        this->SizingCoolOATemp) {
      this->SizingCoolOATemp = state.dataWeatherManager->Environment(env).maxCoolingOATSizing;
      this->SizingCoolOAHumRat = Psychrometrics::PsyWFnTdpPb(
          state,
          state.dataWeatherManager->Environment(env).maxCoolingOADPSizing,
          DataEnvironment::StdPressureSeaLevel);
    }
    if (state.dataWeatherManager->Environment(env).minHeatingOATSizing <
        this->HeatOutTemp) {
      this->HeatOutTemp = state.dataWeatherManager->Environment(env).minHeatingOATSizing;
      this->HeatOutHumRat = Psychrometrics::PsyWFnTdpPb(
          state,
          state.dataWeatherManager->Environment(env).minHeatingOADPSizing,
          DataEnvironment::StdPressureSeaLevel);
    }
  } break;
  case DataGlobalConstants::KindOfSim::RunPeriodWeather: {
  } break;
  default: {
  } break;
  }
}
```

I think this actually looks easiest to read as a range based for loop with a single IF condition, like::

```
for (auto &env : state.dataWeatherManager->Environment) {
  if (env.KindOfEnvrn == DataGlobalConstants::KindOfSim::DesignDay || env.KindOfEnvrn == DataGlobalConstants::KindOfSim::RunPeriodDesign) {
    if (env.maxCoolingOATSizing > this->SizingCoolOATemp) {
      this->SizingCoolOATemp = env.maxCoolingOATSizing;
      this->SizingCoolOAHumRat = Psychrometrics::PsyWFnTdpPb(state, env.maxCoolingOADPSizing, DataEnvironment::StdPressureSeaLevel);
    }
    if (env.minHeatingOATSizing < this->HeatOutTemp) {
      this->HeatOutTemp = env.minHeatingOATSizing;
      this->HeatOutHumRat = Psychrometrics::PsyWFnTdpPb(state, env.minHeatingOADPSizing, DataEnvironment::StdPressureSeaLevel);
    }
  }
}
```

The range-based for loop ensures we only ever loop across the size of the vector, we won't go out of bounds.  In some cases in E+ we obviously don't want to do that, so plain for loops are great.  And like in the other example I comments, the single index is used in multiple arrays in the loop, so that needs to be a regular for loop.  But for cases where you are only traversing a single vector from start to finish, the range-based loops make the code super readable, you don't have to manually create a reference in the loop, and help avoid possible bugs later.

amirroth replied:
`If/else` is good for two cases. One tweak I would make is:

```cpp
if (env.kindOfEnvrn != KindOfSim::DesignDay && env.kindOfEnvrn != KindOfSim::RunPeriodDesign) continue;
```

It reduces the level of logic nesting.  

PS Don't get me starting on naming something `KindOf`. Couldn't you at least have named it `Kinda`?

Myoldmopar commented on tst/EnergyPlus/unit/AirLoopHVACDOAS.unit.cc:
```diff
@@ -4355,111 +4362,22 @@ TEST_F(EnergyPlusFixture, AirLoopHVACDOAS_TestGetDesignDayConditions)
+    EXPECT_NEAR(anotherDOAS.SizingCoolOATemp, state->dataWeatherManager->Environment(2).maxCoolingOATSizing, 0.000001);
+    EXPECT_NEAR(anotherDOAS.SizingCoolOAHumRat, envrnCoolingHumRat, 0.000001);
+    EXPECT_NEAR(anotherDOAS.HeatOutTemp, state->dataWeatherManager->Environment(4).minHeatingOATSizing, 0.000001);
+    EXPECT_NEAR(anotherDOAS.HeatOutHumRat, envrnHeatingHumRat, 0.000001);
```
:+1:

Myoldmopar reviewed (commented):
I'd like to get some thoughts on my comments here, but otherwise I'd assume this is close.  I haven't built with latest develop or verified the issue/fix yet either.

rraustad commented on src/EnergyPlus/WeatherManager.hh:
```diff
@@ -198,6 +198,10 @@ namespace WeatherManager {
+        Real64 maxCoolingOATSizing;      // max outdoor dry-bulb for DesignDay or RunPeriodDesign type weather
+        Real64 maxCoolingOADPSizing;     // max outdoor dew point for DesignDay or RunPeriodDesign type weather
+        Real64 minHeatingOATSizing;      // min outdoor dry-bulb for DesignDay or RunPeriodDesign type weather
+        Real64 minHeatingOADPSizing;     // min outdoor dew point for DesignDay or RunPeriodDesign type weather
```
@Myoldmopar if you are making changes, you could correct the comments on DP. 

    outdoor dew point at max outdoor dry-bulb for DesignDay or RunPeriodDesign type weather
    outdoor dew point at min outdoor dry-bulb for DesignDay or RunPeriodDesign type weather

rraustad commented:
Addressing comments and noticed I used StdPressureSeaLevel then thought about which Pressure to use for moisture conversion. Design days use user input for pressure. Weather data converts using StdBaroPress. Both of these seem reasonable.

    state.dataWeatherManager->TomorrowOutDewPointTemp(ts, hour) =
        Psychrometrics::PsyTdpFnWPb(state, OutHumRat, state.dataWeatherManager->DesDayInput(EnvrnNum).PressBarom);

But design day pressure is only used if that pressure is within 10% of StdBaroPress, otherwise PressBarom = StdBaroPress. So when you see the next commit you'll understand the changes.

    // Check that barometric pressure is within range
    if (state.dataWeatherManager->DesDayInput(EnvrnNum).PressureEntered) {
        if (std::abs((state.dataWeatherManager->DesDayInput(EnvrnNum).PressBarom - state.dataEnvrn->StdBaroPress) /
                     state.dataEnvrn->StdBaroPress) > 0.1) { // 10% off
            ShowWarningError(state,
                     format("SetUpDesignDay: Entered DesignDay Barometric Pressure={:.0R} differs by more than 10% from Standard "
                                "Barometric Pressure={:.0R}.",
                                state.dataWeatherManager->DesDayInput(EnvrnNum).PressBarom,
                                state.dataEnvrn->StdBaroPress));
            ShowContinueError(
                state, "...occurs in DesignDay=" + state.dataEnvrn->EnvironmentName + ", Standard Pressure (based on elevation) will be used.");
            state.dataWeatherManager->DesDayInput(EnvrnNum).PressBarom = state.dataEnvrn->StdBaroPress;
        }
    } else {
        state.dataWeatherManager->DesDayInput(EnvrnNum).PressBarom = state.dataEnvrn->StdBaroPress;
    }

Tested code changes with
[SmallOffice_CentralDOAS_wWeatherDesignDay.idf.txt](https://github.com/NREL/EnergyPlus/files/8910157/SmallOffice_CentralDOAS_wWeatherDesignDay.idf.txt)

rraustad committed:
Address comments and correct moisture conversion

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 9327-DOASSizing-followup

rraustad commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -1018,28 +1018,28 @@ namespace AirLoopHVACDOAS {
+            if (env.maxCoolingOATSizing > this->SizingCoolOATemp) {
+                this->SizingCoolOATemp = env.maxCoolingOATSizing;
+                if (env.KindOfEnvrn == DataGlobalConstants::KindOfSim::DesignDay &&
+                    state.dataWeatherManager->DesDayInput(env.DesignDayNum).PressureEntered) {
```
I verified with a modified example file, SmallOffice_CentralDOAS_wWeatherDesignDay (attached to last PR comment), that (env.DesignDayNum) does = 0 for SizingPeriod:WeatherFileDays (and is non-0 for design days) and that this line does not crash due to a 0 index in DesDayInput(0).PressureEntered.

rraustad commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -1018,28 +1018,28 @@ namespace AirLoopHVACDOAS {
+                    state.dataWeatherManager->DesDayInput(env.DesignDayNum).PressureEntered) {
                     this->SizingCoolOAHumRat = Psychrometrics::PsyWFnTdpPb(
-                        state, state.dataWeatherManager->Environment(env).maxCoolingOADPSizing, DataEnvironment::StdPressureSeaLevel);
+                        state, env.maxCoolingOADPSizing, state.dataWeatherManager->DesDayInput(env.DesignDayNum).PressBarom);
```
switcheroo on pressure used for design day moisture calculation

mjwitte replied:
> switcheroo on pressure used for design day moisture calculation

So, previously, both types of environments were using `DataEnvironment::StdPressureSeaLevel` but now design day environments use `DesDayInput(env.DesignDayNum).PressBarom` and weather environments use `state.dataEnvrn->StdBaroPress`. Why don't they both use `StdBaroPress`? What's used elsewhere for similar calcs?

rraustad replied:
Design days use the outdoor pressure value provided by the user, otherwise that input would not be needed.

    HumidityRatio = Psychrometrics::PsyWFnTdbRhPb(state,
                                        state.dataWeatherManager->TomorrowOutDryBulbTemp(ts, hour),
                                        state.dataWeatherManager->DDHumIndModifier(ts, hour, EnvrnNum) / 100.0,
                                        state.dataWeatherManager->DesDayInput(EnvrnNum).PressBarom);

        Psychrometrics::PsyRhFnTdbWPb(state,
                                      state.dataWeatherManager->TomorrowOutDryBulbTemp(ts, hour),
                                      HumidityRatio,
                                      state.dataWeatherManager->DesDayInput(EnvrnNum).PressBarom,
                                      WeatherManager) * 100.0;

    // TomorrowOutRelHum values set earlier
    state.dataWeatherManager->TomorrowOutDewPointTemp(ts, hour) =
                        Psychrometrics::PsyTdpFnWPb(state, HumidityRatio, state.dataWeatherManager->DesDayInput(EnvrnNum).PressBarom);

mjwitte replied:
Wait, the DOAS sizing is recalculating conditions across design days by looking directly at the design day inputs? Why? Seems it would be better to track peak OA conditions in the sizing routines, since they'll cycle through everything there with HumRat already calculated. Maybe that's for another pass.

And if I'm following `AirLoopDOAS::GetDesignDayConditions` correctly, it only sizes for the peak dry-bulb and coincident humidity with no check on enthalpy? So, a dewpoint design day or other cooler humid day will never be selected?

rraustad replied:
No, the weather manager calculates the max OAT used for sizing and stores that data for each sizing environment. The `AirLoopDOAS::GetDesignDayConditions` uses that data to choose the value used for sizing by searching all design days. There are no sizing routines for air loop systems. This is how it was done before and it was expanded here to include the WeatherfileDays. If there is a better way to choose how to size an outdoor air coil, say based on max enthalpy (which seems reasonable), then that code could be added. Just open an issue and suggest a course of action.

mjwitte replied:
Sorry, I was looking at the code from the prior DOAS sizing PR. Ok, this is clearly an improvement, but I will post an issue to use a better method assuming I can produce an idf that illustrates the problem.  In that issue I'll also suggest that it might be better to track the humidity ratio *after* it's calculated for the current weather in weather manager.

rraustad commented:
Recalling @mjwitte comment [here](https://github.com/NREL/EnergyPlus/pull/9327/files#r827514394) about avoiding design day inputs in calculations. So should we replicate the method for calculating design day dew point or simply use StdBaroPress for the conversion since the design day pressure input will be within 10% of StdBaroPress (i.e., use the switcheroo I just added or delete it?). I already fixed the build warning and will wait on comments before doing anything else.

rraustad committed:
Change loop index type

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 9327-DOASSizing-followup

nrel-bot-3 commented:
@rraustad @myoldmopar it has been 29 days since this pull request was last updated.

Myoldmopar commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2046,6 +2046,29 @@ namespace WeatherManager {
+                    }
+                }
+            }
+
```
I can't remember which comments go on which code blocks, but I still think this could be made much more reasonable with a single reference to an object.  Consider this modification:

```c++
// at the beginning of each day find the min/max weather used for DOAS sizing
auto &thisEnvrn = state.dataWeatherManager->Environment(state.dataWeatherManager->Envrn);
if (thisEnvrn.KindOfEnvrn == DataGlobalConstants::KindOfSim::RunPeriodDesign ||
    thisEnvrn.KindOfEnvrn == DataGlobalConstants::KindOfSim::DesignDay) {
    for (size_t i = 0; i < state.dataWeatherManager->TodayOutDryBulbTemp.size(); ++i) {
        if (state.dataWeatherManager->TodayOutDryBulbTemp[i] > thisEnvrn.maxCoolingOATSizing) {
            thisEnvrn.maxCoolingOATSizing = state.dataWeatherManager->TodayOutDryBulbTemp[i];
            thisEnvrn.maxCoolingOADPSizing = state.dataWeatherManager->TodayOutDewPointTemp[i];
        }
        if (state.dataWeatherManager->TodayOutDryBulbTemp[i] < thisEnvrn.minHeatingOATSizing) {
            thisEnvrn.minHeatingOATSizing = state.dataWeatherManager->TodayOutDryBulbTemp[i];
            thisEnvrn.minHeatingOADPSizing = state.dataWeatherManager->TodayOutDewPointTemp[i];
        }
    }
}
```

amirroth replied:
Don't get me started on this module. Few could use a struct/array refactor more desperately than `WeatherManager`. But yes, shortcut references make code more readable and also debug code run faster too!

Myoldmopar reviewed (commented):
This is probably fine as-is, but I did offer up the code cleanup suggestion again.  If it's totally ready and just needs final testing, I'm happy to drop in the code change when I do local testing ahead of merge.  If there are any other changes needed, it should also still get modified there.  I'll run it locally now, and I'll leave this review as just a 'comment' instead of 'approved' for now.

Myoldmopar commented:
CI is showing the one file with diffs.  I get from your comments above that you took that file and modified it to verify the fix.  Are you also expecting these diffs to show up in the base test file?

rraustad committed:
Update based on comments/suggestions

rraustad commented:
I am adding your suggestion and checking unit tests again.

The one diff is from a change in outdoor pressure when calculating entering coil humidity ratio. Develop (top figure) used sea level pressure and that is not correct since that will result in a humidity ratio that is different from what this coil actually sees. This branch uses pressure from the design day, or if that is not available, uses StdBaroPressure.

![image](https://user-images.githubusercontent.com/5778156/180050038-def5732f-c120-4235-862b-53b8c9b74e28.png)

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 9327-DOASSizing-followup

rraustad committed:
Expand on suggestions

Myoldmopar commented on src/EnergyPlus/WeatherManager.hh:
```diff
@@ -198,6 +198,10 @@ namespace WeatherManager {
+        Real64 maxCoolingOATSizing;      // max outdoor dry-bulb for DesignDay or RunPeriodDesign type weather
+        Real64 maxCoolingOADPSizing;     // outdoor dew point at max outdoor dry-bulb for DesignDay or RunPeriodDesign type weather
+        Real64 minHeatingOATSizing;      // min outdoor dry-bulb for DesignDay or RunPeriodDesign type weather
+        Real64 minHeatingOADPSizing;     // outdoor dew point at min outdoor dry-bulb for DesignDay or RunPeriodDesign type weather
```
When we add new members to classes in the future, just go ahead and initialize them at the declaration:

```
Real64 maxCoolingOATSizing = -999.0;
```

And preferably, not with -999 because that's a magic number.  This is OK for this PR.

Myoldmopar reviewed (approved):
:+1:

Myoldmopar commented:
Linux is still catching up, so I'm inclined to let it work on this one this morning before merging.  But I think it's ready to merge if CI is happy.

Myoldmopar commented:
@rraustad did you see the unit test failure?  It seems some uninitialized index trying to index an array of size 2: `vector::_M_range_check: __n (which is 18446744073709551615) >= this->size() (which is 2)`

I can investigate, but if you happen to already know what it probably is then I'll leave it to you.  I'm going to move on through the PR list for now.  Thanks!

rraustad committed:
Revert too aggressive reference

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 9327-DOASSizing-followup

rraustad commented:
I did not notice the unit test failure. I got a little too agressive with the reference for state.dataWeatherManager->DesDayInput(env.DesignDayNum) since DesignDayNum will be 0 for SizingPeriod:WeatherFileDays.

mjwitte reviewed (commented):
Unit tests succeed, defect file diffs are small'ish.

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2028,24 +2010,37 @@ namespace WeatherManager {
                 }
             }
+
+            // at the beginning of each day find the min/max weather used for DOAS sizing
+            if (thisEnvrn.KindOfEnvrn == DataGlobalConstants::KindOfSim::RunPeriodDesign ||
+                thisEnvrn.KindOfEnvrn == DataGlobalConstants::KindOfSim::DesignDay) {
+                for (size_t i = 0; i < state.dataWeatherManager->TodayOutDryBulbTemp.size(); ++i) {
```
Since the result of this (`thisEnvrn.maxCoolingOATSizing` etc) are only used by the AirloopHVAC DOAS, seems we should set a flag to skip this entirely is there isn't one.

rraustad replied:
Air loops are read in well after this point. The only way I can think of to set a flag is to check for the number of DOAS objects prior to this point. Maybe we need a new function that can scan the json inputs for things like that at the beginning of the simulation. I have wanted to do that for some time.

mjwitte replied:
That's easy enough with 
`bool anyDOAS = (state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, "AirLoopHVAC:DedicatedOutdoorAirSystem") > 0);`
Just a question of a logical spot to put it.

rraustad replied:
@Myoldmopar the comments/suggestion above is where state.dataInputProcessing->inputProcessor->isInputObjectUsed was born.

rraustad committed:
Add bool to new input processor function (only a placeholder)

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 9327-DOASSizing-followup

rraustad commented on src/EnergyPlus/InputProcessing/InputProcessor.cc:
```diff
@@ -466,6 +466,18 @@ bool InputProcessor::checkForUnsupportedObjects(EnergyPlusData &state)
+    // there is no need to process that data if there are no DOAS used in the simulation
+    state.dataInputProcessing->DOASUsedInSim = (state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, "AirLoopHVAC:DedicatedOutdoorAirSystem") > 0);
+}
+
```
To avoid unnecessary code execution in weather manager I added a bool. Chose input processor for the location since that's what this call does. Might want a different location since my expectation is to read in all getInputs prior to sizing. It doesn't matter from an execution time whether getInputs are read early or when the model is called. But from a code order perspective the data needed to process certain objects has to wait on other objects to be read in. An example is UnitarySystem in an air loop where information on zone equipment is needed. If that data were already available the UnitarySystem could process all getInput items in a one-shot. There are other reasons to do this, e.g., if (ZoneEquipmentReadIn) or like statements scattered throughout code. So this is the thought behind this new function,  living here or other location. If this will stay here I will clang format, otherwise that simple change would be a drain on the CI. So what-say-you on this location? and suggestion for other location if needed.

mjwitte replied:
I guess input processor makes sense. I'm not sure about calling this `getSimInputObjects`. That seems a bit premature, especially since it won't be getting an object, just checking how many are present. 

But maybe all of these could be called from a single function along with this new check?
```
        CheckIfAnyEMS(state);
        CheckIfAnyPlant(state);
        CheckIfAnySlabs(state);
        CheckIfAnyBasements(state);
        CheckIfAnyIdealCondEntSetPoint(state);
```

rraustad replied:
Now I think I should have put this new function in Data\EnergyPlusData. Already has all #include needed, is a global access point, and out of the way, and manager, of all other classes. Not for today, but I'm just not sure inputProcessor is the right place.

Myoldmopar replied:
We should be really cautious about putting anything more than the core data classes in the EnergyPlusData world.  I am not saying we shouldn't do it, but we should do it very cautiously.  That was set up in the current way to split things up for clean/fast builds so adding more dependencies there need to be very thought out.

rraustad replied:
That is the same thought I had about InputProcessor. A new module? existing?

Myoldmopar commented:
I think the Clang Format is complaining that the parentheses are unnecessary on src/EnergyPlus/InputProcessing/InputProcessor.cc:478.  Just the one file with diffs otherwise.

rraustad committed:
Address comments and some clean up

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 9327-DOASSizing-followup

mjwitte commented on src/EnergyPlus/InputProcessing/InputProcessor.cc:
```diff
@@ -466,6 +470,24 @@ bool InputProcessor::checkForUnsupportedObjects(EnergyPlusData &state)
     return errorsFound;
 }
 
+void InputProcessor::isInputObjectUsed(EnergyPlusData &state)
```
This works for me, for now.

mjwitte reviewed (approved):
@Myoldmopar Tag! You're it. Merge when ready.

Myoldmopar commented on src/EnergyPlus/InputProcessing/InputProcessor.cc:
```diff
@@ -466,6 +470,24 @@ bool InputProcessor::checkForUnsupportedObjects(EnergyPlusData &state)
+    // there is no need to process that data if there are no DOAS used in the simulation
+    state.dataInputProcessing->AirLoopHVACDOASUsedInSim =
+        state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, "AirLoopHVAC:DedicatedOutdoorAirSystem") > 0;
+    EMSManager::CheckIfAnyEMS(state);
```
So this shouldn't call back _out_ to the EMSManager.  The input processor should only rely on worker routines to develop input structures.  The CheckIfAnyEMS is also a misnamed function, because it doesn't just check if there are any inputs, it assigns a whole bunch of values.  

I need to understand the intent of collecting these, because I could imagine a few ways to make it work better.  I'll ping you separately to discuss it a bit more so we can get this wrapped up.

rraustad replied:
The single intent of this function was to determine if AirLoopHVAC:DedicatedOutdoorAirSystem objects exist in the input. I figure we will need to know if other objects also exist so this function was created for that purpose. It seemed reasonable to also include other "check" functions in SimulationManager here since they are called in the same area of code where this function is called (whether they set a flag or get more data). I don't care if these are moved back to where they were, literaly 3 lines up from this call. I think from just seeing this little example you should have a good idea of what the intent is here, to call for information that would not be available until later in the call order (the AirLoopHVACDOASUsedInSim flag that is, not the other Check functions that were already called here).

![image](https://user-images.githubusercontent.com/5778156/181847116-4d7f9852-0b72-4b61-92cc-ba3bb3a93ea0.png)

rraustad committed:
Move new getFunction to SimulationManager

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 9327-DOASSizing-followup

Myoldmopar commented on src/EnergyPlus/DataGlobals.hh:
```diff
@@ -174,6 +174,7 @@ struct DataGlobal : BaseGlobalStruct
     bool printConsoleOutput = true;
     bool installRootOverride = false;
     int numThread = 1;
+    bool AirLoopHVACDOASUsedInSim = false;
```
We can always keep discussing the best place for these flags, but this is a totally reasonable place to drop it for now.  This flag is intended to be a 'global' flag for a given running simulation.  So DataGlobals is good.

Myoldmopar commented on src/EnergyPlus/SimulationManager.cc:
```diff
@@ -2912,6 +2851,24 @@ namespace SimulationManager {
         state.dataInputProcessing->inputProcessor->preScanReportingVariables(state);
     }
 
+    void isInputObjectUsed(EnergyPlusData &state)
```
Here is good.

Myoldmopar reviewed (approved):
Good to go, thanks for the fix.

Myoldmopar commented:
Thanks @rraustad, let's move forward with this, and we can keep discussing your thoughts on doing early, and intentional, GetInput processing soon, coordinating with the team as a whole too.
